monads/src/HW/Eval.hs

```diff
@@ -26,11 +26,40 @@ data MachineState v = MachineState
 initialState :: MachineState String
 initialState = MachineState [] M.empty
 
--- Execute a single instruction. 
--- Successful evaluation does not produce any useful result: only the effect of modifying state matters. 
+-- Execute a single instruction.
+-- Successful evaluation does not produce any useful result: only the effect of modifying state matters.
 execInstr :: (Ord v, Show v) => StackInstr v -> MyState (MachineState v) (Either (Error v) ())
-execInstr = undefined 
+execInstr (PushNum x) = do
+  MachineState s env <- get
+  put (MachineState (x : s) env)
+  return (Right ())
```

Comment:
Whenever you don't touch parts of the state, it's better to use `modify` instead of `get` and `put`. It's less error-prone: 

```
execInstr (PushNum n) =
  Right <$> modify (\m -> m {getStack = n : getStack m})
```